
def helpMessage() {
    log.info nfcoreHeader()
    log.info"""

    Usage:

    The typical command for running the pipeline is as follows:

    nextflow run nf-core/methylseq --reads '*_R{1,2}.fastq.gz' -profile docker

    Mandatory arguments:
      --aligner                 Alignment tool to use (default: bismark)
                                        Available: bismark, bismark_hisat, bwameth
      --reads                   Path to input data (must be surrounded with quotes)
      -profile                  Configuration profile to use. Can use multiple (comma separated)
                                        Available: conda, docker, singularity, awsbatch, test and more.

    Options:
     --genome    Name of iGenomes reference
     --singleEnd       Specifies that the input is single end reads
     --comprehensive        Output information for all cytosine contexts
     --ignoreFlags      Run MethylDackel with the flag to ignore SAM flags.
     --meth_cutoff      Specify a minimum read coverage to report a methylation call during Bismark's bismark_methylation_extractor step.
     --mindepth     Specify a minimum read coverage for MethylDackel to report a methylation call.
     --methylKit        Run MethylDackel with the --methylKit flag to produce files suitable for use with the methylKit R package.
     --nodedup      Skip deduplication step after alignment. This is turned on automatically if --rrbs is specified
     --non_directional      Run alignment against all four possible strands
     --saveAlignedIntermediates     Save aligned intermediates to results directory
     --saveTrimmed      Save trimmed reads to results directory
     --unmapped     Save unmapped reads to fastq files
     --relaxMismatches      Turn on to relax stringency for alignment (set allowed penalty with --numMismatches)
     --numMismatches        0.6 will allow a penalty of bp * -0.6 - for 100bp reads (bismark default is 0.2)
     --known_splices	Supply a .gtf file containing known splice sites (bismark_hisat only)
     --slamseq	Run bismark in SLAM-seq mode

    References                      If not specified in the configuration file or you wish to overwrite any of the references.
      --fasta                       Path to Fasta reference
      --fasta_index           Path to Fasta Index
      --bismark_index      Path to Bismark index
      --bwa_meth_index  Path to bwameth index
      --saveReference       Save reference(s) to results directory

    Trimming options:
     --notrim   Skip read trimming
     --clip_r1  Trim the specified number of bases from the 5' end of read 1 (or single-end reads).
     --clip_r2  Trim the specified number of bases from the 5' end of read 2 (paired-end only).
     --three_prime_clip_r1 Trim the specified number of bases from the 3' end of read 1 AFTER adapter/quality trimming
     --three_prime_clip_r2 Trim the specified number of bases from the 3' end of read 2 AFTER adapter/quality trimming
     --rrbs Turn on if dealing with MspI digested material.

    Trimming presets:
     --pbat
     --single_cell
     --epignome
     --accell
     --zymo
     --cegx

    Other options:
      --outdir                      The output directory where the results will be saved
      --email                       Set this parameter to your e-mail address to get a summary e-mail with details of the run sent to you when the workflow exits
      --maxMultiqcEmailFileSize     Theshold size for MultiQC report to be attached in notification email. If file generated by pipeline exceeds the threshold, it will not be attached (Default: 25MB)
      -name                         Name for the pipeline run. If not specified, Nextflow will automatically generate a random mnemonic.

    AWSBatch options:
      --awsqueue                    The AWSBatch JobQueue that needs to be set when running on AWSBatch
      --awsregion                   The AWS Region for your AWS Batch job to run on
    """.stripIndent()
}
